### DISCUSSION TITLE
Compile issue (missing __app_stack_top/_bottom) with Bare Metal Application as HSS Payload Example (LIM-Debug)

### DISCUSSION THREAD

**Discussion by ChrRudel** [polarfire-soc/polarfire-soc-documentation#109] [Bare metal embedded software] [answered]
The instructions on the following page specify to set IMAGE_LOADED_BY_BOOTLOADER to 1.

https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/software-development/hss-payloads.md#config-bare-metal-payload

"Ensure that IMAGE_LOADED_BY_BOOTLOADER define is set to 1."

With this change to the "mss_sw_config.h" I am running in the following compile error when trying to compile the LIM-Debug configuration of the "mpfs-mmuart-interrupt" example (2021.11 release of the polarfire-soc-bare-metal-examples).

c:/microchip/softconsole-v2021.3-7.0.0.599/riscv-unknown-elf-gcc/bin/../lib/gcc/riscv64-unknown-elf/8.3.0/../../../../riscv64-unknown-elf/bin/ld.exe: ./src/platform/mpfs_hal/startup_gcc/mss_entry.o: in function `.L0 ':
C:\tmp\Workspace\mpfs-mmuart-interrupt\LIM-Debug/../src/platform/mpfs_hal/startup_gcc/mss_entry.S:350: undefined reference to `__app_stack_bottom'
c:/microchip/softconsole-v2021.3-7.0.0.599/riscv-unknown-elf-gcc/bin/../lib/gcc/riscv64-unknown-elf/8.3.0/../../../../riscv64-unknown-elf/bin/ld.exe: C:\tmp\Workspace\mpfs-mmuart-interrupt\LIM-Debug/../src/platform/mpfs_hal/startup_gcc/mss_entry.S:351: undefined reference to `__app_stack_top'
c:/microchip/softconsole-v2021.3-7.0.0.599/riscv-unknown-elf-gcc/bin/../lib/gcc/riscv64-unknown-elf/8.3.0/../../../../riscv64-unknown-elf/bin/ld.exe: C:\tmp\Workspace\mpfs-mmuart-interrupt\LIM-Debug/../src/platform/mpfs_hal/startup_gcc/mss_entry.S:352: undefined reference to `__app_stack_top'

It seems that the required __app_stack_bottom & _app_stack_top symbols (used in mss_entry.s for the IMAGE_LODED_BY_BOOTLOADER ===1 case) are not defined in the "mpfs-lim.ld" Linker script.

friendly regards,
Christian

**Comment by hughbreslin** [comment, marked as answer]
Hi @ChrRudel 

I had a quick look into this one - its due to the fact that we now assume any bare metal payload loaded by the HSS is going to:

1. target a single hart - we are releasing an update to support applications that are going to target multiple harts in an upcoming release
2. Execute from the DDR

I can reproduce the issue you're seeing by building for LIM-Debug or LIM-Release in the included build configurations but building using the DDR release configuration succeeds. 

Could you try the following:
1. Make a copy of the "mpfs-ddr-loaded-by-boot-loader.ld" linker script and name it "mpfs-l2lim-loaded-by-boot-loader.ld"
2. Modify sections that point to the DDR (`> ddr_cached_32bit`) to point to LIM instead (`> l2lim`) 
3. Save this file and create a new build configuration called "L2LIM-Release-loaded-by-bootloader" that copies the settings of the DDR-Release build configuration:
![image](https://user-images.githubusercontent.com/44862881/176697476-3bd974d7-aa92-4214-b00c-f32a122df894.png)
4. Modify the build configuration to point to the new linker script that was created in step 1:
![image](https://user-images.githubusercontent.com/44862881/176721661-d7e21ad0-c37e-4d31-9b94-0bb810442d95.png)
5. Build using the "L2LIM-Release-loaded-by-bootloader" build configuration that was created

You should now be able to build the project for LIM release.

**Reply by ChrRudel** [reply]
Thanks @hughbreslin, your suggested workaround (create "L2LIM-Release-loaded-by-bootloader" configuration) worked for me. With the suggested steps, I can compile the project again.
friendly regards,
Christian

**Reply by hughbreslin** [reply]
Excellent, it looks like the documentation could do with an update so I’ll get that done and we can discuss adding a lim loaded by boot loader linker script :)
